File: built-in plugins dsp/mb-mix-transformer.cpp
/**
 * MB Transformer
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Audio transformer emulation for harmonic warmth
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_TRANSFORMER_H
#define MB_MIX_TRANSFORMER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixTransformer : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-transformer";
    static constexpr const char* PLUGIN_NAME    = "MB Transformer";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.3f;  // range [0, 1]
    float impedance = 0.5f;  // range [0, 1]
    float lowEnd = 0.5f;  // range [0, 1]
    float output = 0.8f;  // range [0, 1]
    };

    MbMixTransformer() = default;
    ~MbMixTransformer() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.impedance = std::clamp(params.impedance, 0f, 1f);
        params.lowEnd = std::clamp(params.lowEnd, 0f, 1f);
        params.output = std::clamp(params.output, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Transformer
        return input;
    }
};

#endif // MB_MIX_TRANSFORMER_H

File: built-in plugins dsp/mb-mix-transformer.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixTransformerPlugin: PluginDefinition = { id: 'mb-mix-transformer', slug: 'mb-mix-transformer', name: 'MB Transformer', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'Audio transformer emulation for harmonic warmth', author: 'Max Booster', parameters: [{ id: 'drive', name: 'Drive', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'impedance', name: 'Impedance', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'lowEnd', name: 'Low End', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'output', name: 'Output', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { drive: 0.3, impedance: 0.5, lowEnd: 0.5, output: 0.8 } };

export default MbMixTransformerPlugin;
